ENH: add windows support

RubyLiu206 opened the pull request:
Windows support was added mostly by:

- Fixing path separator issues
- Making sure test executables existed within a regular windows install
- Adding support python programs if this wasn't the case.

Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #**:

**Description of the changes being introduced by the pull request**:

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

alyptik commented on in_toto/models/link.py:
```diff
@@ -109,6 +109,10 @@ def type_(self):
   def read(data):
     """Static method to instantiate a new Link from a Python dictionary """
+
+    #Todo: add some fitures to develop
```
Small spelling error; probably meant to be "features."

alyptik commented on in_toto/models/metadata.py:
```diff
@@ -107,13 +108,14 @@ def load(path):
     signed_data = data.get("signed", {})
     signed_type = signed_data.get("_type")
-
+    #Todo:add some fitures to develop
```
Same spellng error.

SantiagoTorres replied:
I'd probably drop these changes from this PR

trishankatdatadog replied:
What are these missing features?

SantiagoTorres replied:
This is part of another pr, #215 . I think these markers were put in there when we were discussing what features to add..

alyptik commented on in_toto/models/metadata.py:
```diff
@@ -56,6 +56,7 @@ def __repr__(self):
           "signatures": self.signatures,
           "signed": attr.asdict(self.signed)
+          #Todo:add some fitures to develop
         }, indent=1, separators=(",", ": "), sort_keys=True)
```
Same spellng error.

trishankatdatadog replied:
What are these missing features?

alyptik commented on in_toto/models/metadata.py:
```diff
@@ -46,7 +46,7 @@ class Metablock(ValidationMixin):
     self.signatures = kwargs.get("signatures", [])
     self.signed = kwargs.get("signed")
-
+    #Todo:add some fitures here to develop
```
Same spellng error.

trishankatdatadog replied:
What are these missing features?

alyptik commented on in_toto/models/metadata.py:
```diff
@@ -107,13 +108,14 @@ def load(path):
     elif signed_type == "layout":
       signed = Layout.read(signed_data)
 
+    # Todo: add some fitures to develop with the
```
Same spellng error.

trishankatdatadog replied:
What are these missing features?

SantiagoTorres replied:
For this and all of the above. We need to remove these comments as they are not relevant to this changeset.

alyptik commented on in_toto/verifylib.py:
```diff
@@ -680,7 +680,7 @@ def verify_match_rule(rule, source_artifacts_queue, source_artifacts, links):
     filtered_source_paths = []
     # Add trailing slash to source prefix if it does not exist
-    normalized_source_prefix = os.path.join(rule_data["source_prefix"], "")
+    normalized_source_prefix = os.path.join(rule_data["source_prefix"], )
```
Is there a reason for the extra comma here?

SantiagoTorres replied:
os.path.join needs two params, but passing an empty string doesn't work like it should in certain python versions...

alyptik replied:
Ah, that's a strange quirk, noted.

alyptik commented on tests/test_in_toto_verify.py:
```diff
@@ -80,6 +83,7 @@ def setUpClass(self):
     # Copy demo files to temp dir
     for file in os.listdir(demo_files):
       shutil.copy(os.path.join(demo_files, file), self.test_dir)
+    shutil.copytree(scripts_directory, 'scripts')
```
Was the indentation level change here intended?

SantiagoTorres replied:
is this na indentation change or just a *new* copytree statement. I think this is right

alyptik replied:
Yeah I figured it is possibly intended, but I don't have enough context to be sure, so just figured I'd mention it.

alyptik reviewed (changes requested):
Aside from a few spelling errors and small nitpicks, looks good to me.

trishankatdatadog commented:
@SantiagoTorres @alyptik How can we help push these changes ASAP?

We are happy test and address the comments in this PR. @ofek

trishankatdatadog commented:
I think we should add AppVeyor to make sure that in-toto works on Windows. We don't have to do it right now with this PR, but we should add testing against Windows.

trishankatdatadog commented on in_toto/models/link.py:
```diff
@@ -109,6 +109,10 @@ def type_(self):
     """Static method to instantiate a new Link from a Python dictionary """
+
+    #Todo: add some fitures to develop
+    #Todo: Material/product/environment
```
What is there TODO here?

SantiagoTorres replied:
Same as above...

trishankatdatadog commented on in_toto/runlib.py:
```diff
@@ -297,6 +300,7 @@ def execute_link(link_cmd_args, record_streams):
     - The return value of the executed command.
   """
   # TODO: Properly duplicate standard streams (issue #11)
+  print("hello Ruby")
```
This needs to be removed.

trishankatdatadog commented on in_toto/runlib.py:
```diff
@@ -436,6 +440,7 @@ def in_toto_run(name, material_list, product_list, link_cmd_args,
     Newly created Metablock object containing a Link object
 
   """
+  print("hello ruby")
```
This needs to be removed.

trishankatdatadog commented on tests/common.py:
```diff
@@ -74,6 +74,18 @@ def assert_cli_sys_exit(self, cli_args, status):
-
-    self.assertEqual(raise_ctx.exception.code, status)
+      if raise_ctx.exception.code != status:
+         import pdb;pdb.set_trace()
```
This needs to be removed.

trishankatdatadog commented on tests/common.py:
```diff
@@ -74,6 +74,18 @@ def assert_cli_sys_exit(self, cli_args, status):
     """
     with patch.object(sys, "argv", [self.file_name]
         + cli_args), self.assertRaises(SystemExit) as raise_ctx:
+      print("about to run with: {}".format(sys.argv))
```
This needs to be removed.

trishankatdatadog commented on in_toto/runlib.py:
```diff
@@ -251,9 +251,12 @@ def record_artifacts_as_dict(artifacts, exclude_patterns=None,
         if exclude_patterns:
           filepaths = _apply_exclude_patterns(filepaths, exclude_patterns)
+        # FIXME: this is necessary to provide consisency between windows filepaths and
+        # *nix filepaths. A better solution may be in order though...
```
I am not convinced that this PR would fix all issues on Windows.

For example, are there checks to prevent absolute paths such as `C://Program Files/in-toto/run.py`?

SantiagoTorres commented:
@RubyLiu206 there are a lot of stray comments from our discussion of PR #215 

@trishankatdatadog the appveyor aspect was added on #207, but we haven't worked on it until this PR is merged.

RubyLiu206 commented:
I will delete those today.

On Tue, Sep 18, 2018, 10:56 AM Santiago Torres <notifications@github.com>
wrote:

> @RubyLiu206 <https://github.com/RubyLiu206> there are a lot of stray
> comments from our discussion of PR #215
> <https://github.com/in-toto/in-toto/pull/215>
>
> @trishankatdatadog <https://github.com/trishankatdatadog> the appveyor
> aspect was added on #207 <https://github.com/in-toto/in-toto/pull/207>,
> but we haven't worked on it until this PR is merged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/in-toto/in-toto/pull/211#issuecomment-422427076>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AovNdLfECbEjqtNhjye7G_7RZA8pGcyMks5ucQmfgaJpZM4WdWyd>
> .
>

trishankatdatadog commented:
Okay, thanks, everyone 🙇 

We'd like to get this package, please, before releasing a new version of tool (cc @ofek)

trishankatdatadog commented:
I just realized looking at an [example of a link metadata file produced on Windows](https://github.com/DataDog/integrations-core/blob/16e5789d22b742f49d391ead2a4886d69ca89e83/.links/tag.c295cf63.link) that `in-toto` should accept Windows path separators, even if it was somehow written and signed.

One way to do this would be to internally normalize all path separators when reading metadata.

Are these security reasons why this would be a bad idea? What do you guys think?

SantiagoTorres commented:
@trishankatdatadog let me see if I understand. you're suggesting to not normalize the paths upon writing, but rather normalize them upon reading the metadata? (probably upon verification)

I don't think there's anything wrong with that. Do you see any benefits for doing it though?

trishankatdatadog commented:
@SantiagoTorres Sorry, I wasn't clear. Let's do both :) So let's normalize when writing, but _also_ when reading. I was just worried about being too unforgiving with the unexpected inputs (unless there were good security reasons, of course).

SantiagoTorres commented:
We could do that, yes. I don't see anything against that. We'd have to make sure that both path separators are invalid characters in all platforms, which I think is true.

RubyLiu206 committed:
ENH: add windows support

SantiagoTorres committed:
FIX: gpg: functions: remove unused var 'errors'

SantiagoTorres committed:
FIX: test: in_toto_{verify,run}: fix script setup

The scripts directory wasn't being copied over when running the tests,
which lead to a failure on non-windows systems.

SantiagoTorres commented:
@alyptik I dismissed your review since this PR was a complete rewrite in the end. Could you still review and ticketize anything I may have missed? Thanks!

lukpueh commented on tests/scripts/expr:
```diff
@@ -0,0 +1,29 @@
+
+  <Date>
+    1527623020
+"""
```
I think a doc-copy-paste mistake happened here.
